# DATA CLEANING & PREPARATION

* This file is for cleaning all three hree data sets , and prepreing them for the EDA and the modeling phase

In [273]:
# import all necessary libraries

import pandas as pd

pd.set_option('display.max_rows', None)


In [196]:
# load all data sets first to avoid multiple reads of the same file

tmdb_data = pd.read_csv("Data sets/TMDB_dataset.csv")
movies_data = pd.read_csv("Data sets/movies.csv")
ratings_data = pd.read_csv("Data sets/ratings.csv")


## 1. Datasets Quick Overview

In [197]:

movies_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   movieId  9742 non-null   int64 
 1   title    9742 non-null   object
 2   genres   9742 non-null   object
dtypes: int64(1), object(2)
memory usage: 228.5+ KB


In [198]:
ratings_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   userId     100836 non-null  int64  
 1   movieId    100836 non-null  int64  
 2   rating     100836 non-null  float64
 3   timestamp  100836 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 3.1 MB


In [199]:
# check the structure of the TMDB dataset
tmdb_data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1408433 entries, 0 to 1408432
Data columns (total 24 columns):
 #   Column                Non-Null Count    Dtype  
---  ------                --------------    -----  
 0   id                    1408433 non-null  int64  
 1   title                 1408415 non-null  object 
 2   vote_average          1408433 non-null  float64
 3   vote_count            1408433 non-null  int64  
 4   status                1408433 non-null  object 
 5   release_date          1098259 non-null  object 
 6   revenue               1408433 non-null  int64  
 7   runtime               1408433 non-null  int64  
 8   adult                 1408433 non-null  bool   
 9   backdrop_path         350692 non-null   object 
 10  budget                1408433 non-null  int64  
 11  homepage              145028 non-null   object 
 12  imdb_id               665867 non-null   object 
 13  original_language     1408433 non-null  object 
 14  original_title        1408415 non-

In [200]:
# check columns for all datasets

print("columns for the tmdb dataset " , tmdb_data.columns)
print("columns for the movies dataset " , movies_data.columns)
print("columns for the ratings dataset " , ratings_data.columns)

columns for the tmdb dataset  Index(['id', 'title', 'vote_average', 'vote_count', 'status', 'release_date',
       'revenue', 'runtime', 'adult', 'backdrop_path', 'budget', 'homepage',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'tagline', 'genres',
       'production_companies', 'production_countries', 'spoken_languages',
       'keywords'],
      dtype='object')
columns for the movies dataset  Index(['movieId', 'title', 'genres'], dtype='object')
columns for the ratings dataset  Index(['userId', 'movieId', 'rating', 'timestamp'], dtype='object')


In [201]:
tmdb_data.head()  

,id,title,vote_average,vote_count,status,release_date,revenue,runtime,adult,backdrop_path,...,original_title,overview,popularity,poster_path,tagline,genres,production_companies,production_countries,spoken_languages,keywords
0,27205,Inception,8.364,34495,Released,2010-07-15,825532764,148,False,/8ZTVqvKDQ8emSGUEMjsS4yHAwrp.jpg,...,Inception,"Cobb, a skilled thief who commits corporate es...",83.952,/oYuLEt3zVCKq57qu2F8dT7NIa6f.jpg,Your mind is the scene of the crime.,"Action, Science Fiction, Adventure","Legendary Pictures, Syncopy, Warner Bros. Pict...","United Kingdom, United States of America","English, French, Japanese, Swahili","rescue, mission, dream, airplane, paris, franc..."
1,157336,Interstellar,8.417,32571,Released,2014-11-05,701729206,169,False,/pbrkL804c8yAv3zBZR4QPEafpAR.jpg,...,Interstellar,The adventures of a group of explorers who mak...,140.241,/gEU2QniE6E77NI6lCU6MxlNBvIx.jpg,Mankind was born on Earth. It was never meant ...,"Adventure, Drama, Science Fiction","Legendary Pictures, Syncopy, Lynda Obst Produc...","United Kingdom, United States of America",English,"rescue, future, spacecraft, race against time,..."
2,155,The Dark Knight,8.512,30619,Released,2008-07-16,1004558444,152,False,/nMKdUUepR0i5zn0y1T4CsSB5chy.jpg,...,The Dark Knight,Batman raises the stakes in his war on crime. ...,130.643,/qJ2tW6WMUDux911r6m7haRef0WH.jpg,Welcome to a world without rules.,"Drama, Action, Crime, Thriller","DC Comics, Legendary Pictures, Syncopy, Isobel...","United Kingdom, United States of America","English, Mandarin","joker, sadism, chaos, secret identity, crime f..."
3,19995,Avatar,7.573,29815,Released,2009-12-15,2923706026,162,False,/vL5LR6WdxWPjLPFRLe133jXWsh5.jpg,...,Avatar,"In the 22nd century, a paraplegic Marine is di...",79.932,/kyeqWdyUXW608qlYkRqosgbbJyK.jpg,Enter the world of Pandora.,"Action, Adventure, Fantasy, Science Fiction","Dune Entertainment, Lightstorm Entertainment, ...","United States of America, United Kingdom","English, Spanish","future, society, culture clash, space travel, ..."
4,24428,The Avengers,7.710,29166,Released,2012-04-25,1518815515,143,False,/9BBTo63ANSmhC4e6r62OJFuK2GL.jpg,...,The Avengers,When an unexpected enemy emerges and threatens...,98.082,/RYMX2wcKCBAr24UyPD7xwmjaTn.jpg,Some assembly required.,"Science Fiction, Action, Adventure",Marvel Studios,United States of America,"English, Hindi, Russian","new york city, superhero, shield, based on com..."


In [202]:
movies_data.head()

,movieId,title,genres
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
1,2,Jumanji (1995),Adventure|Children|Fantasy
2,3,Grumpier Old Men (1995),Comedy|Romance
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance
4,5,Father of the Bride Part II (1995),Comedy


In [203]:
ratings_data.head()

,userId,movieId,rating,timestamp
0,1,1,4.0,964982703
1,1,3,4.0,964981247
2,1,6,4.0,964982224
3,1,47,5.0,964983815
4,1,50,5.0,964982931


In [204]:
ratings_data.value_counts("userId")

userId
414    2698
599    2478
474    2108
448    1864
274    1346
       ... 
442      20
278      20
147      20
320      20
53       20
Name: count, Length: 610, dtype: int64

In [205]:
# check the movie ids in the all datasets to see if they match and can be used for merging

print("unique movie ids in tmdb dataset " , tmdb_data["id"].nunique())
print("unique movie ids in movies dataset " , movies_data["movieId"].nunique())
print("unique movie ids in ratings dataset " , ratings_data["movieId"].nunique())   

unique movie ids in tmdb dataset  1407200
unique movie ids in movies dataset  9742
unique movie ids in ratings dataset  9724


### Summary on the data sets overview

* well the movie ids in the movies and ratings dataset match but they do not match with the tmdb dataset. We will have to merge the tmdb dataset with the movies dataset first and then merge the resulting dataset with the ratings dataset.
* This would mean alot of the movies in the tmdb dataset are not in the movies dataset and ratings dataset. 
* We will have to drop the tmdb dataset for our analysis as we will not be able to merge it with the other two datasets.

## 2. Schema Standardization

Here we'll be doing the making the datasets:

* consistent
* predictable
* merge-ready

### 2.1 Normlize column names

In [206]:
def normalize_columns(df):
    df.columns = (
        df.columns
        .str.strip()
        .str.lower()
        .str.replace(" ", "_")
    )
    return df

movies_data = normalize_columns(movies_data)
ratings_data = normalize_columns(ratings_data)
tmdb_data = normalize_columns(tmdb_data)

# check the column names again to confirm they are normalized
print("columns for the tmdb dataset " , tmdb_data.columns)
print("columns for the movies dataset " , movies_data.columns)
print("columns for the ratings dataset " , ratings_data.columns)

columns for the tmdb dataset  Index(['id', 'title', 'vote_average', 'vote_count', 'status', 'release_date',
       'revenue', 'runtime', 'adult', 'backdrop_path', 'budget', 'homepage',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'tagline', 'genres',
       'production_companies', 'production_countries', 'spoken_languages',
       'keywords'],
      dtype='object')
columns for the movies dataset  Index(['movieid', 'title', 'genres'], dtype='object')
columns for the ratings dataset  Index(['userid', 'movieid', 'rating', 'timestamp'], dtype='object')


### 2.2 Enforce Data Types

In [207]:
ratings_data = ratings_data.astype({
    "userid": "int32",
    "movieid": "int32",
    "rating": "float32"
})

movies_data["movieid"] = movies_data["movieid"].astype("int32")
tmdb_data["id"] = tmdb_data["id"].astype("int32")


# check the data types again to confirm they are correct
print("data types for the tmdb dataset " , tmdb_data[["id"]].dtypes)
print("data types for the movies dataset " , movies_data[["movieid"]].dtypes)
print("data types for the ratings dataset " , ratings_data[["userid", "movieid", "rating"]].dtypes)

data types for the tmdb dataset 

 id    int32
dtype: object
data types for the movies dataset  movieid    int32
dtype: object
data types for the ratings dataset  userid       int32
movieid      int32
rating     float32
dtype: object


### 2.3 Timestamp ?Handling

In [208]:
ratings_data["timestamp"] = pd.to_datetime(ratings_data["timestamp"], unit="s")

# check the data types again to confirm the timestamp is converted correctly
print("data types for the ratings dataset " , ratings_data[["timestamp"]].dtypes)

data types for the ratings dataset  timestamp    datetime64[ns]
dtype: object


### 2.4 Drop Unwanted Features(noise)

In [209]:
# drop the following features in tmdb backdrop_path
"""
poster_path
homepage
status
production_companies
production_countries
spoken_languages

"""

tmdb_data_dropped = tmdb_data.drop(columns=[
    "backdrop_path",
    "poster_path",
    "homepage",
    "status",
    "production_companies",
    "production_countries",
    "spoken_languages" 
])

In [210]:
#check the structure of the TMDB dataset after dropping the columns
tmdb_data_dropped.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1408433 entries, 0 to 1408432
Data columns (total 17 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   id                 1408433 non-null  int32  
 1   title              1408415 non-null  object 
 2   vote_average       1408433 non-null  float64
 3   vote_count         1408433 non-null  int64  
 4   release_date       1098259 non-null  object 
 5   revenue            1408433 non-null  int64  
 6   runtime            1408433 non-null  int64  
 7   adult              1408433 non-null  bool   
 8   budget             1408433 non-null  int64  
 9   imdb_id            665867 non-null   object 
 10  original_language  1408433 non-null  object 
 11  original_title     1408415 non-null  object 
 12  overview           1086361 non-null  object 
 13  popularity         1408433 non-null  float64
 14  tagline            196047 non-null   object 
 15  genres             791155 non-nu

## 3. Clean the Linking Fields

### 3.1 Create Clean Titles 

In [211]:
import re # for regular expressions

def clean_title(title):
    if pd.isna(title):
        return ""
    
    title = title.lower()
    title = re.sub(r"\(\d{4}\)", "", title)   # remove year
    title = re.sub(r"[^a-z0-9 ]", "", title)  # remove punctuation
    title = re.sub(r"\s+", " ", title)        # normalize spaces
    return title.strip()

movies_data["clean_title"] = movies_data["title"].apply(clean_title)
tmdb_data_dropped["clean_title"] = tmdb_data_dropped["title"].apply(clean_title)

In [212]:
# check the titles after cleaning to confirm they are cleaned correctly
print("cleaned titles in the tmdb dataset " , tmdb_data_dropped["clean_title"].head())
print("cleaned titles in the movies dataset " , movies_data["clean_title"].head())

cleaned titles in the tmdb dataset  0          inception
1       interstellar
2    the dark knight
3             avatar
4       the avengers
Name: clean_title, dtype: object
cleaned titles in the movies dataset  0                      toy story
1                        jumanji
2               grumpier old men
3              waiting to exhale
4    father of the bride part ii
Name: clean_title, dtype: object


In [213]:
print("colums after " , movies_data.columns)
print("colums after " , tmdb_data_dropped.columns)

colums after  Index(['movieid', 'title', 'genres', 'clean_title'], dtype='object')
colums after  Index(['id', 'title', 'vote_average', 'vote_count', 'release_date', 'revenue',
       'runtime', 'adult', 'budget', 'imdb_id', 'original_language',
       'original_title', 'overview', 'popularity', 'tagline', 'genres',
       'keywords', 'clean_title'],
      dtype='object')


### 3.2 Handle Duplicates

In [214]:
#check the duplicates in the tmdb dataset
tmdb_data_dropped.duplicated().sum()

np.int64(388)

In [215]:
#check the duplicates in the movies dataset
movies_data.duplicated().sum()

np.int64(0)

In [216]:
# check the duplicates in the ratings dataset
ratings_data.duplicated().sum()

np.int64(0)

In [217]:
# drop duplicates in the tmdb dataset
tmdb_data_dropped = tmdb_data_dropped.drop_duplicates()

#check if dropped duplicates in the tmdb dataset
print("duplicates in the tmdb dataset after dropping " , tmdb_data_dropped.duplicated().sum())

duplicates in the tmdb dataset after dropping  0


## 4. Merge Strategy

* Merging the Tmdb dataset and the movies dataset first (based on title)
* Then the result with the movie ratings (based on ratings)

### Step 4.1: Movies + TMDB

In [218]:
movie_metadata = movies_data.merge(
    tmdb_data_dropped,
    on="clean_title",
    how="inner"
)
# This ensures that those movies that are present in both datasets are retained, while those that do not have a match in either dataset are excluded.

In [219]:
#check the structure of the merged dataset
movie_metadata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33338 entries, 0 to 33337
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   movieid            33338 non-null  int32  
 1   title_x            33338 non-null  object 
 2   genres_x           33338 non-null  object 
 3   clean_title        33338 non-null  object 
 4   id                 33338 non-null  int32  
 5   title_y            33338 non-null  object 
 6   vote_average       33338 non-null  float64
 7   vote_count         33338 non-null  int64  
 8   release_date       27309 non-null  object 
 9   revenue            33338 non-null  int64  
 10  runtime            33338 non-null  int64  
 11  adult              33338 non-null  bool   
 12  budget             33338 non-null  int64  
 13  imdb_id            20998 non-null  object 
 14  original_language  33338 non-null  object 
 15  original_title     33338 non-null  object 
 16  overview           297

In [220]:
movie_metadata.head()

,movieid,title_x,genres_x,clean_title,id,title_y,vote_average,vote_count,release_date,revenue,...,adult,budget,imdb_id,original_language,original_title,overview,popularity,tagline,genres_y,keywords
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy,toy story,862,Toy Story,7.971,17152,1995-10-30,394400000,...,False,30000000,tt0114709,en,Toy Story,"Led by Woody, Andy's toys live happily in his ...",78.404,Hang on for the comedy that goes to infinity a...,"Animation, Adventure, Family, Comedy","rescue, friendship, mission, martial arts, jea..."
1,2,Jumanji (1995),Adventure|Children|Fantasy,jumanji,8844,Jumanji,7.239,9833,1995-12-15,262821940,...,False,65000000,tt0113497,en,Jumanji,When siblings Judy and Peter discover an encha...,13.444,Roll the dice and unleash the excitement!,"Adventure, Fantasy, Family","giant insect, board game, disappearance, jungl..."
2,3,Grumpier Old Men (1995),Comedy|Romance,grumpier old men,15602,Grumpier Old Men,6.476,347,1995-12-22,71500000,...,False,25000000,tt0113228,en,Grumpier Old Men,A family wedding reignites the ancient feud be...,14.815,Still Yelling. Still Fighting. Still Ready for...,"Romance, Comedy","fishing, sequel, old man, best friend, wedding..."
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance,waiting to exhale,31357,Waiting to Exhale,6.183,142,1995-12-22,81452156,...,False,16000000,tt0114885,en,Waiting to Exhale,"Cheated on, mistreated and stepped on, the wom...",14.451,Friends are the people who let you be yourself...,"Comedy, Drama, Romance","based on novel or book, interracial relationsh..."
4,5,Father of the Bride Part II (1995),Comedy,father of the bride part ii,11862,Father of the Bride Part II,6.228,659,1995-12-08,76594107,...,False,0,tt0113041,en,Father of the Bride Part II,Just when George Banks has recovered from his ...,14.537,Just when his world is back to normal... he's ...,"Comedy, Family","daughter, baby, parent child relationship, mid..."


#### $Summary from  merging 
  * Only 33338 entries were kept

### 4.2: Merge with Ratings


In [221]:
# merge the ratings dataset with the movie metadata dataset to get the final dataset for analysis
final_dataset = ratings_data.merge(
    movie_metadata,
    on="movieid",
    how="inner"
)
print("structure of the final dataset " , final_dataset.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 321026 entries, 0 to 321025
Data columns (total 24 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   userid             321026 non-null  int32         
 1   movieid            321026 non-null  int32         
 2   rating             321026 non-null  float32       
 3   timestamp          321026 non-null  datetime64[ns]
 4   title_x            321026 non-null  object        
 5   genres_x           321026 non-null  object        
 6   clean_title        321026 non-null  object        
 7   id                 321026 non-null  int32         
 8   title_y            321026 non-null  object        
 9   vote_average       321026 non-null  float64       
 10  vote_count         321026 non-null  int64         
 11  release_date       262143 non-null  object        
 12  revenue            321026 non-null  int64         
 13  runtime            321026 non-null  int64   

 ### 4.3 Check the structure of the movies dataset

In [222]:
final_dataset.head()

,userid,movieid,rating,timestamp,title_x,genres_x,clean_title,id,title_y,vote_average,...,adult,budget,imdb_id,original_language,original_title,overview,popularity,tagline,genres_y,keywords
0,1,1,4.0,2000-07-30 18:45:03,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy,toy story,862,Toy Story,7.971,...,False,30000000,tt0114709,en,Toy Story,"Led by Woody, Andy's toys live happily in his ...",78.404,Hang on for the comedy that goes to infinity a...,"Animation, Adventure, Family, Comedy","rescue, friendship, mission, martial arts, jea..."
1,1,3,4.0,2000-07-30 18:20:47,Grumpier Old Men (1995),Comedy|Romance,grumpier old men,15602,Grumpier Old Men,6.476,...,False,25000000,tt0113228,en,Grumpier Old Men,A family wedding reignites the ancient feud be...,14.815,Still Yelling. Still Fighting. Still Ready for...,"Romance, Comedy","fishing, sequel, old man, best friend, wedding..."
2,1,6,4.0,2000-07-30 18:37:04,Heat (1995),Action|Crime|Thriller,heat,949,Heat,7.903,...,False,60000000,tt0113277,en,Heat,Obsessive master thief Neil McCauley leads a t...,41.522,A Los Angeles crime saga.,"Action, Crime, Drama","robbery, chase, obsession, detective, heist, t..."
3,1,6,4.0,2000-07-30 18:37:04,Heat (1995),Action|Crime|Thriller,heat,42023,Heat,5.900,...,False,0,tt0093164,en,Heat,Reynolds plays an ex-soldier-of-fortunish char...,3.913,NaN,"Action, Thriller",NaN
4,1,6,4.0,2000-07-30 18:37:04,Heat (1995),Action|Crime|Thriller,heat,4579,Heat,5.979,...,False,0,tt0068688,en,Heat,"Former child star Joe Davis, reduced to living...",2.521,NaN,"Comedy, Drama","ladykiller, diva, seduction, satire, older wom..."


In [223]:
final_dataset.isnull().sum()

userid                    0
movieid                   0
rating                    0
timestamp                 0
title_x                   0
genres_x                  0
clean_title               0
id                        0
title_y                   0
vote_average              0
vote_count                0
release_date          58883
revenue                   0
runtime                   0
adult                     0
budget                    0
imdb_id              123545
original_language         0
original_title            0
overview              38730
popularity                0
tagline              203265
genres_y             100873
keywords             179112
dtype: int64

In [224]:
final_dataset.duplicated().sum()

np.int64(0)

#### $Summary of the final merging 

 * Only 321026 entries were gotten from now the three data sets 
 * no duplicated data
 * over 3 of needed features have null values (we'll be dealt with below)

In [225]:
non_empty = final_dataset[final_dataset["keywords"] != ""]
print(len(non_empty))

321026


## 5.Post Merge Cleaning  

### 5.1 Handle Individual columns

In [226]:
final_dataset.isnull().sum()

userid                    0
movieid                   0
rating                    0
timestamp                 0
title_x                   0
genres_x                  0
clean_title               0
id                        0
title_y                   0
vote_average              0
vote_count                0
release_date          58883
revenue                   0
runtime                   0
adult                     0
budget                    0
imdb_id              123545
original_language         0
original_title            0
overview              38730
popularity                0
tagline              203265
genres_y             100873
keywords             179112
dtype: int64

### Priority Order 

You don’t treat all missing data equally.

1. Tier 1 (critical for content model):
  → overview,keywords

2. Tier 2 (nice-to-have):
  → tagline

3. Tier 3 (ignore for now):
→ imdb_id, release_date , genres_y

### a) Handle Overview 

* This is the primary source of data for our analysis (NLP), and it is crucial to ensure that it is clean and free of duplicates or missing values before we proceed with any analysis or modeling.
* we just fill with null vals since TF-IDF can handle null values. 
* It is better to keep the data as it is rather than dropping rows with missing values, which could lead to loss of valuable information.

In [227]:
final_dataset["overview"].isnull().sum()    

np.int64(38730)

In [228]:
final_dataset["overview"] = final_dataset["overview"].fillna("")

In [229]:
#check the structure of the overview column after filling null values
print("null values in the overview column after filling " , final_dataset["overview"].isnull().sum())

null values in the overview column after filling  0


### b) Handle Keywords

In [230]:
non_empty = final_dataset[final_dataset["keywords"] != ""]
print(len(non_empty))

321026


In [231]:
# fill misssing values in the keywords column with empty string

#final_dataset["keywords"] = final_dataset["keywords"].fillna("")

In [232]:
final_dataset["keywords"].isnull().sum()

np.int64(179112)

 extract the keywords from the keywords column aand clean them to get a list of keywords for each movie

import ast # for converting string representation of list to actual list

def extract_keywords(x):
    try:
        items = ast.literal_eval(x)
        return " ".join([i["name"] for i in items])
    except:
        return ""

final_dataset["keywords"] = final_dataset["keywords"].apply(extract_keywords)


In [233]:
print(final_dataset["keywords"].value_counts().head(10))

keywords
short film                                                                                                                                                                                                                                                                                                                                                                                                                                                          3348
gay pornography                                                                                                                                                                                                                                                                                                                                                                                                                                                     1940
woman director                                                               

In [234]:
#check the keywords column after extracting keywords to confirm they are extracted correctly
# 1. Inspect the specific range you mentioned
print("Keywords subset (23000-23005):")
print(final_dataset["keywords"].loc[23000:23005])

# 2. Filter for rows where the string has more than 3 words
# We split by whitespace and count the resulting list elements
long_keywords = final_dataset[final_dataset["keywords"].str.split().str.len() > 3]

# 3. View the results
print("\nRows with more than 3 words in 'keywords':")
print(long_keywords["keywords"].head())

Keywords subset (23000-23005):
23000                                                  NaN
23001                                                  NaN
23002                                                  NaN
23003                                                  NaN
23004    drug dealer, corruption, drug abuse, unsociabi...
23005    drug dealer, amnesia, insulin, tattoo, waitres...
Name: keywords, dtype: object

Rows with more than 3 words in 'keywords':
0    rescue, friendship, mission, martial arts, jea...
1    fishing, sequel, old man, best friend, wedding...
2    robbery, chase, obsession, detective, heist, t...
4    ladykiller, diva, seduction, satire, older wom...
7                        working class, woman director
Name: keywords, dtype: object


### c) Handle genres_y

In [235]:
final_dataset["genres"] = final_dataset["genres_y"].fillna(final_dataset["genres_x"])


final_dataset["genres"].isnull().sum()

np.int64(0)

In [236]:
final_dataset.columns 

Index(['userid', 'movieid', 'rating', 'timestamp', 'title_x', 'genres_x',
       'clean_title', 'id', 'title_y', 'vote_average', 'vote_count',
       'release_date', 'revenue', 'runtime', 'adult', 'budget', 'imdb_id',
       'original_language', 'original_title', 'overview', 'popularity',
       'tagline', 'genres_y', 'keywords', 'genres'],
      dtype='object')

In [237]:
final_dataset["genres"].value_counts()

genres
Drama                                                35949
Comedy                                               16391
Documentary                                          15666
Animation                                             8982
Horror                                                7484
                                                     ...  
Science Fiction, Thriller, Mystery, Drama                1
Crime, Horror, Fantasy, Science Fiction                  1
Action, Horror, Science Fiction, Thriller, Comedy        1
Fantasy, Animation, Science Fiction, Family              1
Drama, Fantasy, Horror, Mystery                          1
Name: count, Length: 2885, dtype: int64

* drop the two genres with final genre being used

In [238]:
#drop both genre_x and genre_y since we have the final genre column
final_dataset = final_dataset.drop(columns=["genres_x", "genres_y"])

In [239]:
# confirm the final structure of the dataset after cleaning
print("final structure of the dataset after removing the two genres " , final_dataset.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 321026 entries, 0 to 321025
Data columns (total 23 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   userid             321026 non-null  int32         
 1   movieid            321026 non-null  int32         
 2   rating             321026 non-null  float32       
 3   timestamp          321026 non-null  datetime64[ns]
 4   title_x            321026 non-null  object        
 5   clean_title        321026 non-null  object        
 6   id                 321026 non-null  int32         
 7   title_y            321026 non-null  object        
 8   vote_average       321026 non-null  float64       
 9   vote_count         321026 non-null  int64         
 10  release_date       262143 non-null  object        
 11  revenue            321026 non-null  int64         
 12  runtime            321026 non-null  int64         
 13  adult              321026 non-null  bool    

### d) Handling tagline



In [240]:
final_dataset.isna().sum()

userid                    0
movieid                   0
rating                    0
timestamp                 0
title_x                   0
clean_title               0
id                        0
title_y                   0
vote_average              0
vote_count                0
release_date          58883
revenue                   0
runtime                   0
adult                     0
budget                    0
imdb_id              123545
original_language         0
original_title            0
overview                  0
popularity                0
tagline              203265
keywords             179112
genres                    0
dtype: int64

In [241]:
final_dataset["tagline"] = final_dataset["tagline"].fillna("")

#check the structure of the tagline column after filling null values
print(final_dataset["tagline"].isnull().sum())

0


### e) Handling the Release year 



In [242]:
final_dataset["release_year"] = pd.to_datetime(
    final_dataset["release_date"], errors="coerce"
).dt.year

# check the structure of the release_year column after extracting the year to confirm it is extracted correctly
print(final_dataset["release_year"].head())
print(final_dataset["release_year"].isnull().sum())

0    1995.0
1    1995.0
2    1995.0
3    1986.0
4    1972.0
Name: release_year, dtype: float64
58883


In [243]:
# handle null values and also convert to string type for the content column which will be used for the NLP analysis
final_dataset["release_year"] = final_dataset["release_year"].fillna(0).astype(int).astype(str)

In [244]:
# check the structure of the release year column after handling null values and converting to string type to confirm it is done correctly
print("release year column after handling null values and converting to string type " , final_dataset["release_year"].isnull().sum())
print(final_dataset["release_year"].astype)

release year column after handling null values and converting to string type  0
<bound method NDFrame.astype of 0         1995
1         1995
2         1995
3         1986
4         1972
          ... 
321021    2010
321022    2026
321023       0
321024       0
321025    2017
Name: release_year, Length: 321026, dtype: object>


### f) Drop the IMDB ID 

In [245]:
final_dataset.columns

Index(['userid', 'movieid', 'rating', 'timestamp', 'title_x', 'clean_title',
       'id', 'title_y', 'vote_average', 'vote_count', 'release_date',
       'revenue', 'runtime', 'adult', 'budget', 'imdb_id', 'original_language',
       'original_title', 'overview', 'popularity', 'tagline', 'keywords',
       'genres', 'release_year'],
      dtype='object')

In [246]:
#final_dataset = final_dataset.drop(columns=["imdb_id"] , axis=1)

print("final structure of the dataset after dropping the imdb_id column " , final_dataset.columns)

final structure of the dataset after dropping the imdb_id column  Index(['userid', 'movieid', 'rating', 'timestamp', 'title_x', 'clean_title',
       'id', 'title_y', 'vote_average', 'vote_count', 'release_date',
       'revenue', 'runtime', 'adult', 'budget', 'imdb_id', 'original_language',
       'original_title', 'overview', 'popularity', 'tagline', 'keywords',
       'genres', 'release_year'],
      dtype='object')


### g) Handle the duplicate titles 

In [247]:
# print both titles 

final_dataset[["title_x", "title_y"]].head()

,title_x,title_y
0,Toy Story (1995),Toy Story
1,Grumpier Old Men (1995),Grumpier Old Men
2,Heat (1995),Heat
3,Heat (1995),Heat
4,Heat (1995),Heat


In [248]:
#drop the redundant columns , we'll use the cleaned title column for analysis and drop the original title columns
final_dataset = final_dataset.drop(columns=["title_x", "title_y"])

In [249]:
# confirm the final structure of the dataset after dropping the original title columns
print("final structure of the dataset after dropping the original title columns " , final_dataset.info())
final_dataset.columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 321026 entries, 0 to 321025
Data columns (total 22 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   userid             321026 non-null  int32         
 1   movieid            321026 non-null  int32         
 2   rating             321026 non-null  float32       
 3   timestamp          321026 non-null  datetime64[ns]
 4   clean_title        321026 non-null  object        
 5   id                 321026 non-null  int32         
 6   vote_average       321026 non-null  float64       
 7   vote_count         321026 non-null  int64         
 8   release_date       262143 non-null  object        
 9   revenue            321026 non-null  int64         
 10  runtime            321026 non-null  int64         
 11  adult              321026 non-null  bool          
 12  budget             321026 non-null  int64         
 13  imdb_id            197481 non-null  object  

Index(['userid', 'movieid', 'rating', 'timestamp', 'clean_title', 'id',
       'vote_average', 'vote_count', 'release_date', 'revenue', 'runtime',
       'adult', 'budget', 'imdb_id', 'original_language', 'original_title',
       'overview', 'popularity', 'tagline', 'keywords', 'genres',
       'release_year'],
      dtype='object')

### h) Handle adult column

In [250]:
final_dataset["adult"].value_counts()

adult
False    302667
True      18359
Name: count, dtype: int64

In [251]:
final_dataset["adult"].isnull().sum()

np.int64(0)

### i) Handle runtime 

In [252]:
final_dataset["runtime"].isnull().sum()

np.int64(0)

In [259]:
final_dataset["runtime"].astype

<bound method NDFrame.astype of 0          81
1         101
2         170
3         101
4         102
         ... 
321021     94
321022      0
321023      0
321024      0
321025    136
Name: runtime, Length: 321026, dtype: int64>

### j) Handle Original language 

In [274]:
# check all languages 
final_dataset["original_language"].value_counts()

original_language
en    223485
fr     10002
zh      9839
ru      7502
es      7388
ja      6873
de      6453
pt      4128
ko      4075
it      3805
fa      2448
hi      2073
xx      1881
ar      1867
tr      1802
pl      1736
cn      1681
cs      1626
nl      1563
el      1159
ta      1151
id      1020
tl      1001
sv       976
ml       945
uk       837
bn       824
no       795
th       757
he       756
da       709
hu       683
et       570
bg       533
kn       525
mk       503
fi       461
sh       400
te       365
sr       358
vi       352
hy       333
hr       309
ms       293
mn       293
kk       290
ka       287
lv       287
sk       266
sq       234
az       234
is       234
sl       233
ro       220
ca       219
lt       152
nb       112
kw        82
bs        80
pa        71
eu        70
si        64
ur        53
zu        49
uz        45
my        45
ne        42
ab        39
mr        37
km        37
se        34
as        33
yi        30
or        28
ky        28
ku     

In [ ]:
# Apply mapping according to the iso standard to convert the language codes to language names


language_map = {
    "en": "English",
    "fr": "French",
    "zh": "Chinese",
    "ru": "Russian",
    "es": "Spanish",
    "ja": "Japanese",
    "de": "German",
    "pt": "Portuguese",
    "ko": "Korean",
    "it": "Italian",
    "fa": "Persian",
    "hi": "Hindi",
    "xx": "No Language",
    "ar": "Arabic",
    "tr": "Turkish",
    "pl": "Polish",
    "cn": "Cantonese/Chinese",
    "cs": "Czech",
    "nl": "Dutch",
    "el": "Greek",
    "ta": "Tamil",
    "id": "Indonesian",
    "tl": "Tagalog",
    "sv": "Swedish",
    "os": "Ossetian",
    "dv": "Divehi",
    "wo": "Wolof",
    "nv": "Navajo",
    "mi": "Maori"
}

# Apply the mapping to your dataset
final_dataset['language_name'] = final_dataset['original_language'].map(language_map)

In [277]:
#verify the new language_name column is created correctly

final_dataset["language_name"].value_counts()

language_name
English              223485
French                10002
Chinese                9839
Russian                7502
Spanish                7388
Japanese               6873
German                 6453
Portuguese             4128
Korean                 4075
Italian                3805
Persian                2448
Hindi                  2073
No Language            1881
Arabic                 1867
Turkish                1802
Polish                 1736
Cantonese/Chinese      1681
Czech                  1626
Dutch                  1563
Greek                  1159
Tamil                  1151
Indonesian             1020
Tagalog                1001
Swedish                 976
Ossetian                  1
Divehi                    1
Wolof                     1
Navajo                    1
Maori                     1
Name: count, dtype: int64

### Check the finaldataset after handling individual columns 

In [253]:
final_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 321026 entries, 0 to 321025
Data columns (total 22 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   userid             321026 non-null  int32         
 1   movieid            321026 non-null  int32         
 2   rating             321026 non-null  float32       
 3   timestamp          321026 non-null  datetime64[ns]
 4   clean_title        321026 non-null  object        
 5   id                 321026 non-null  int32         
 6   vote_average       321026 non-null  float64       
 7   vote_count         321026 non-null  int64         
 8   release_date       262143 non-null  object        
 9   revenue            321026 non-null  int64         
 10  runtime            321026 non-null  int64         
 11  adult              321026 non-null  bool          
 12  budget             321026 non-null  int64         
 13  imdb_id            197481 non-null  object  

## 5.2 Build the Final Content Field

* we combine the following features into a single content feature for text analysis overview, keywords, genres and tagline
* The features include overview, keywords, genres and tagline which are all text features that can be used for text analysis and building a content based recommendation system.
* use this content feature for text analysis and building a content based recommendation system.
   * It Also helps with the nissing values in the individual features as we are combining them into a single content feature.
   * This way we can still use the information from the non-null values in the individual features even if some of them have null values.

In [278]:
# create a content column by combining the following columns which will be used for the NLP analysis
cols = [
    "overview",
    "keywords",
    "genres",
    "tagline",
    "language_name",
    "release_year",
    "runtime"
]

final_dataset["content"] = (
    final_dataset[cols]
    .fillna("")
    .astype(str)
    .agg(" ".join, axis=1)
)

In [279]:
# view the content column after creating it to confirm it is created correctly

final_dataset["content"].head(20)

0     Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy's heart, Woody plots against Buzz. But when circumstances separate Buzz and Woody from their owner, the duo eventually learns to put aside their differences. rescue, friendship, mission, martial arts, jealousy, bullying, elementary school, rivalry, anthropomorphism, friends, computer animation, buddy, walkie talkie, toy car, boy next door, new toy, neighborhood, toy comes to life, resourcefulness Animation, Adventure, Family, Comedy Hang on for the comedy that goes to infinity and beyond! English 1995 81
1                                                                                                                                         A family wedding reignites the ancient feud between next-door neighbors and fishing buddies John and Max. Meanwhile, a sultry Italian divorcée opens a restaurant at the local bait shop, alarming the l

In [280]:
final_dataset["content"].str.len().describe()

count    321026.000000
mean        324.426003
std         234.151800
min          13.000000
25%         129.000000
50%         277.000000
75%         486.000000
max        1413.000000
Name: content, dtype: float64

In [281]:
import pandas as pd

pd.set_option("display.max_colwidth", None)

print(final_dataset["content"].iloc[0])

Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy's heart, Woody plots against Buzz. But when circumstances separate Buzz and Woody from their owner, the duo eventually learns to put aside their differences. rescue, friendship, mission, martial arts, jealousy, bullying, elementary school, rivalry, anthropomorphism, friends, computer animation, buddy, walkie talkie, toy car, boy next door, new toy, neighborhood, toy comes to life, resourcefulness Animation, Adventure, Family, Comedy Hang on for the comedy that goes to infinity and beyond! English 1995 81


In [282]:
final_dataset.head()

,userid,movieid,rating,timestamp,clean_title,id,vote_average,vote_count,release_date,revenue,...,original_language,original_title,overview,popularity,tagline,keywords,genres,release_year,content,language_name
0,1,1,4.0,2000-07-30 18:45:03,toy story,862,7.971,17152,1995-10-30,394400000,...,en,Toy Story,"Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy's heart, Woody plots against Buzz. But when circumstances separate Buzz and Woody from their owner, the duo eventually learns to put aside their differences.",78.404,Hang on for the comedy that goes to infinity and beyond!,"rescue, friendship, mission, martial arts, jealousy, bullying, elementary school, rivalry, anthropomorphism, friends, computer animation, buddy, walkie talkie, toy car, boy next door, new toy, neighborhood, toy comes to life, resourcefulness","Animation, Adventure, Family, Comedy",1995,"Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy's heart, Woody plots against Buzz. But when circumstances separate Buzz and Woody from their owner, the duo eventually learns to put aside their differences. rescue, friendship, mission, martial arts, jealousy, bullying, elementary school, rivalry, anthropomorphism, friends, computer animation, buddy, walkie talkie, toy car, boy next door, new toy, neighborhood, toy comes to life, resourcefulness Animation, Adventure, Family, Comedy Hang on for the comedy that goes to infinity and beyond! English 1995 81",English
1,1,3,4.0,2000-07-30 18:20:47,grumpier old men,15602,6.476,347,1995-12-22,71500000,...,en,Grumpier Old Men,"A family wedding reignites the ancient feud between next-door neighbors and fishing buddies John and Max. Meanwhile, a sultry Italian divorcée opens a restaurant at the local bait shop, alarming the locals who worry she'll scare the fish away. But she's less interested in seafood than she is in cooking up a hot time with Max.",14.815,Still Yelling. Still Fighting. Still Ready for Love.,"fishing, sequel, old man, best friend, wedding, italian restaurant, old friends, duringcreditsstinger, pranks","Romance, Comedy",1995,"A family wedding reignites the ancient feud between next-door neighbors and fishing buddies John and Max. Meanwhile, a sultry Italian divorcée opens a restaurant at the local bait shop, alarming the locals who worry she'll scare the fish away. But she's less interested in seafood than she is in cooking up a hot time with Max. fishing, sequel, old man, best friend, wedding, italian restaurant, old friends, duringcreditsstinger, pranks Romance, Comedy Still Yelling. Still Fighting. Still Ready for Love. English 1995 101",English
2,1,6,4.0,2000-07-30 18:37:04,heat,949,7.903,6552,1995-12-15,187436818,...,en,Heat,Obsessive master thief Neil McCauley leads a top-notch crew on various daring heists throughout Los Angeles while determined detective Vincent Hanna pursues him without rest. Each man recognizes and respects the ability and the dedication of the other even though they are aware their cat-and-mouse game may end in violence.,41.522,A Los Angeles crime saga.,"robbery, chase, obsession, detective, heist, thief, honor, murder, betrayal, gang, los angeles, california, cat and mouse, bank robbery, criminal mastermind, ex-con, one last job, loner, bank job, neo-noir, crime epic","Action, Crime, Drama",1995,"Obsessive master thief Neil McCauley leads a top-notch crew on various daring heists throughout Los Angeles while determined detective Vincent Hanna pursues him without rest. Each man recognizes and respects the ability and the dedication of the other even though they are aware their cat-and-mouse game may end in violence. robbery, chase, obsession, detective, heist, thief, honor, murder, betrayal, gang, los angeles, california, cat and mouse, bank robbery, criminal mastermind, ex-con, one last job, loner, bank 

In [283]:
final_dataset.columns

Index(['userid', 'movieid', 'rating', 'timestamp', 'clean_title', 'id',
       'vote_average', 'vote_count', 'release_date', 'revenue', 'runtime',
       'adult', 'budget', 'imdb_id', 'original_language', 'original_title',
       'overview', 'popularity', 'tagline', 'keywords', 'genres',
       'release_year', 'content', 'language_name'],
      dtype='object')

In [284]:
final_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 321026 entries, 0 to 321025
Data columns (total 24 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   userid             321026 non-null  int32         
 1   movieid            321026 non-null  int32         
 2   rating             321026 non-null  float32       
 3   timestamp          321026 non-null  datetime64[ns]
 4   clean_title        321026 non-null  object        
 5   id                 321026 non-null  int32         
 6   vote_average       321026 non-null  float64       
 7   vote_count         321026 non-null  int64         
 8   release_date       262143 non-null  object        
 9   revenue            321026 non-null  int64         
 10  runtime            321026 non-null  int64         
 11  adult              321026 non-null  bool          
 12  budget             321026 non-null  int64         
 13  imdb_id            197481 non-null  object  

## REMARKS AFTER CLEANING 

* WHAT TO KEEP (FEATURES)
* HANDLING THE KEYWORDS MISSING VALUES 
* MAKE TWO DATASETS WITH NEEDED FEATURES FOR BOTH MODELS 